StashApiClient: Run HTTP requests in the main thread

proski opened the pull request:
```
*  StashApiClient: Run HTTP requests in the main thread
   
   Using separate threads adds complexity for little benefit. Socket timeout
   should take care of slow HTTP connections.
   
   JENKINS-30558 doesn't provide any evidence that a separate timeout for
   requests was needed. The request timeout and the socket timeout were added
   in the same commit 80431f422dce9ea763e59c28e35fef75113203b1.
   
   This change eliminates ExecutionException as an exception thrown by
   StashApiClient, replacing it with more specific exceptions. Update unit
   tests for the new behavior.
```

This PR depends on #143.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:
```diff
@@ -213,60 +205,29 @@ private String getRequest(String path) throws StashApiException {
+    String response = "";
     try {
       request.addHeader(new BasicScheme().authenticate(credentials, request, null));
+      HttpResponse httpResponse = client.execute(request);
```
it seems this block is mostly the same for `get` an `post`, is there a chance to do some code reuse?

proski replied:
Yes, but not in this PR. My first priority is correctness. Code reuse will be done once it's clear which code we need. For instance, POST requests can look at the HTTP status code code or at the response contents or at both. I will need to decide what is best by looking at the API specification. Then I will see if there is a pattern that can be refactored.

I may actually need to go in the opposite direction at one point. `validResponseCode` is misleading, as different requests expect different status codes. `409 Conflict` is unquestionably valid for the merge requests. `validResponseCode` conflates two issues, valid/invalid and success/error.

proski replied:
Actually, the HttpResponse contents processing can be refactored. I did it as the first commit, which reduced the amount of code being moved in the larger second commit. I also decided to throw on empty response contents, as we expect JSON, and an empty string is not valid JSON anyway.

proski commented:
Added a commit to refactor processing of the HTTP response contents.

Empty HTTP response throws when JSON is expected.

proski commented:
Changed the first commit to use EntityUtils. Using UTF-8 unconditionally as in the original code, it's a requirement of JSON.

proski commented:
Moved the first commit to #143 to simplify the review.

proski committed:
StashApiClient: Run HTTP requests in the main thread

Using separate threads adds complexity for little benefit. Socket timeout
should take care of slow HTTP connections.

JENKINS-30558 doesn't provide any evidence that a separate timeout for
requests was needed. The request timeout and the socket timeout were added
in the same commit 80431f422dce9ea763e59c28e35fef75113203b1.

This change eliminates ExecutionException as an exception thrown by
StashApiClient, replacing it with more specific exceptions. Update unit
tests for the new behavior.

proski commented:
Rebased on top of the updated #143.

proski committed:
StashApiClient: Declare httpTask and thread where they are initialized

proski commented:
* Updated on top or the updated #143.
* Split out a commit that makes the threading code more compact. The whole purpose it to simplify the review.
* Reverted some unnecessary changes to simplify review.

proski commented:
Rebased and ready to go. No more dependencies of other PRs.
